cocos/core/geometry/spline.ts

```diff
@@ -0,0 +1,292 @@
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ */
+
+/**
+ * @packageDocumentation
+ * @module geometry
+ */
+
+import { assertIsTrue } from '../data/utils/asserts';
+import { clamp, Vec3 } from '../math';
+import { warnID } from '../platform/debug';
+import enums from './enums';
+
+export enum SplineMode {
+    /**
+     * Broken line:
+     * Each knot is connected with a straight line from the beginning to the end to form a curve. At least two knots.
+     */
+    LINEAR = 0,
+
+    /**
+     * Piecewise Bezier curve:
+     * Every four knots form a curve. Total knots number must be a multiple of 4.
+     * Each curve passes only the first and fourth knots, and does not pass through the middle two control knots.
+     * 
+     * If you need a whole continuous curve:
+     * (1) Suppose the four knots of the previous curve are A, B, C, D
+     * (2) The four knots of the next curve must be D, E, F, G
+     * (3) C and E need to be symmetrical about D
+     */
+    BEZIER = 1,
+
+    /**
+     * Catmull Rom curve:
+     * All knots(including start & end knots) form a whole continuous curve. At least two knots.
+     * The whole curve passes through all knots.
+     */
+    CATMULL_ROM = 2
+};
+
+const SPLINE_WHOLE_INDEX = 0xffffffff;
+
+const _v0 = new Vec3();
+const _v1 = new Vec3();
+const _v2 = new Vec3();
+const _v3 = new Vec3();
+
+/**
+ * @en
+ * Basic Geometry: Spline.
+ * @zh
+ * 基础几何 Spline
+ */
+
+export class Spline {
+    private readonly _type: number;
+    private _mode: SplineMode = SplineMode.CATMULL_ROM;
+    private _knots: Vec3[] = [];
+
+    public constructor(mode: SplineMode = SplineMode.CATMULL_ROM, knots: Vec3[] = []) {
+        this._type = enums.SHAPE_SPLINE;
+        this._mode = mode;
+
+        for (let i = 0; i < knots.length; i++) {
+            this._knots[i] = new Vec3(knots[i]);
+        }
+    }
+
+    public static create (mode: SplineMode, knots: Vec3[] = []) {
+        return new Spline(mode, knots);
+    }
+
+    public static clone (s: Spline) {
+        return new Spline(s.getMode(), s.getKnots());
+    }
+
+    public static copy (out: Spline, s: Spline) {
+        out._mode = s.getMode();
+        out._knots = [];
+
+        for (let i = 0; i < s.getKnots().length; i++) {
+            out._knots[i] = new Vec3(s.getKnots()[i]);
+        }
+
+        return out;
+    }
+
+    get type () {
+        return this._type;
+    }
+
+    public getMode () {
+        return this._mode;
+    }
+
+    public setMode (value) {
+        this._mode = value;
+    }
+
+    public getKnots () {
+        return this._knots;
+    }
+
+    public setKnots (value) {
+        this._knots = [];
```

Comment:
out._knots already exists, create a new one is not needed. May be can just do out._knots.length = 0 to reuse the array.